Web Inspector: Create All Fonts Tab in Font Section

francescorn opened the pull request:
#### b6f332bf9b5fc5c853b02114cc98cdf1bd489272
<pre>
Web Inspector: Create All Fonts Tab in Font Section
<a href="https://bugs.webkit.org/show_bug.cgi?id=260127">https://bugs.webkit.org/show_bug.cgi?id=260127</a>

Reviewed by NOBODY (OOPS!).

Co-authored by Brandon Stewart.

We created a Custom Fonts tab to display all of the Custom Fonts in the DOM, and to auto refresh to account for adjusting webpages such as slideshows.
There are parameters font-family name, stretch, style, and loaded, and can sort by each parameter as well.
If there are no custom fonts in the DOM, &quot;No Custom Fonts&quot; will be displayed.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getCustomFonts):
* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles):
(WI.DOMNodeStyles.prototype.get customFonts):
(WI.DOMNodeStyles.prototype.refresh):
* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel.prototype.update):
(WI.FontDetailsPanel.prototype.initialLayout):
(WI.FontDetailsPanel.prototype._sortDataGrid.comparator):
(WI.FontDetailsPanel.prototype._sortDataGrid):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b6f332bf9b5fc5c853b02114cc98cdf1bd489272

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54582 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34001 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7154 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57863 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5316 "Hash b6f332bf for PR 29525 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56884 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41540 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5318 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/57863 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/59/builds/5316 "Hash b6f332bf for PR 29525 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56677 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/41540 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/55/builds/7154 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/57863 "Hash b6f332bf for PR 29525 does not build (failure)") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/41540 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/55/builds/7154 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3456 "Hash b6f332bf for PR 29525 does not build (failure)") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/47947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/41540 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/55/builds/7154 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59453 "Hash b6f332bf for PR 29525 does not build (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/54088 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/29815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/61/builds/5318 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/59453 "Hash b6f332bf for PR 29525 does not build (failure)") | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30965 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/55/builds/7154 "Hash b6f332bf for PR 29525 does not build (failure)") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/59453 "Hash b6f332bf for PR 29525 does not build (failure)") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/31943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/66386 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/30753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12656 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
+    auto customFonts = JSON::ArrayOf<Inspector::Protocol::CSS::FontFace>::create();
+    auto* domAgent = m_instrumentingAgents.persistentDOMAgent();
+
+    for (auto* document : domAgent->documents()) {
```
there's no guarantee that the `domAgent` will exist here, so id either wrap this entire `for` in a `if (auto* domAgent = m_instrumentingAgents.persistentDOMAgent()) { ... }` or add a `return` here in a `if (!domAgent)`

francescorn replied:
I replaced with this:
```
if (auto* domAgent = m_instrumentingAgents.persistentDOMAgent()) {
        for (auto* document : domAgent->documents()) {
```

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
+    auto* domAgent = m_instrumentingAgents.persistentDOMAgent();
+
+    for (auto* document : domAgent->documents()) {
+        auto iterator = document->fonts().get().createIterator(document->scriptExecutionContext());
```
you can do `document->fonts()->` instead of calling `.get()`

francescorn replied:
I removed the `.get`s

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
+        auto iterator = document->fonts().get().createIterator(document->scriptExecutionContext());
+
+        while (auto font = iterator.next()) {
+
```
Style: unnecessary newline

francescorn replied:
I removed the new line

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
+
+        while (auto font = iterator.next()) {
+
+            String status = ""_s;
```
no need to assign to `""_s` as `status` will always be set below

francescorn replied:
I removed status here.

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
+                break;
+            default:
+                ASSERT_NOT_REACHED();
+                break;
```
this is unnecessary since you've already covered all the possible values of `FontFace::LoadStatus`

francescorn replied:
I removed the assert.

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
+
+            String status = ""_s;
+
+            switch (font.get()->status()) {
```
you can do `font->` instead of calling `.get()`

francescorn replied:
I removed .get from everything I added it.

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
+            .setSizeAdjust(font.get()->sizeAdjust())
+            .setLoaded(status)
+            .release();
+            customFonts->addItem(result);
```
this should be `customFonts->addItem(WTFMove(result))` since it's no longer needed (i.e. we can relinquish ownership)

francescorn replied:
I replaced ```customFonts->addItem(result);``` with ```customFonts->addItem(WTFMove(result));```

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
+            }
+
+            Ref<Inspector::Protocol::CSS::FontFace> result = Inspector::Protocol::CSS::FontFace::create()
+            .setFamily(font.get()->family())
```
you can do `font->` instead of calling `.get()`

Style: also please indent all of these
```cpp
auto fontFace = Inspector::Protocol::CSS::FontFace::create()
    .setFamily(font->family())
    .setWeight(font->weight())
    ...
    .release();
customFonts->addItem(WTFMove(fontFace));
```

dcrousso commented on Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
```diff
@@ -156,6 +156,10 @@ localizedStrings["All Changes"] = "All Changes";
 /* Break (pause) on all exceptions */
 localizedStrings["All Exceptions @ JavaScript Breakpoint"] = "All Exceptions";
+/* Font Details Panel */
+localizedStrings["No Custom Fonts @ Font Details Sidebar Property"] = "No Custom Fonts";
```
please do not manually edit this file.  it's autogenerated by running
```
./Tools/Scripts/extract-localizable-js-strings --utf8 ./Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js ./Source/WebInspectorUI/UserInterface
```

francescorn replied:
I ran the command and got:
```/* Message shown when there are no custom fonts. */
localizedStrings["No Custom Fonts @ Font Details Sidebar Property"] = "No Custom Fonts";```
populated.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -135,6 +137,8 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
     get usedCSSVariables() { return this._usedCSSVariables; }
     get allCSSVariables() { return this._allCSSVariables; }
 
```
Style: unnecessary newline

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -340,6 +344,15 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
 
+
+        if (InspectorBackend.hasCommand("CSS.getCustomFonts")) {
+            target.CSSAgent.getCustomFonts((error, fontFamilyNames) =>{
```
calling this every time we create/update a `WI.DOMNodeStyles` seems like way too much since it gets all fonts for the entire page, not just for a specific node (which is really what `WI.DOMNodeStyles` is about

id suggest moving this to `WI.CSSManager` and having an event from the backend (e.g. `CSS.fontsChanged`) that then triggers this to be called again, as that way we dont call it more than once for the same data

Style: missing space `=> {`

dcrousso commented on Source/JavaScriptCore/inspector/protocol/CSS.json:
```diff
@@ -462,6 +477,14 @@
         },
+        {
+            "name": "getCustomFonts",
+            "description": "Returns all loaded fonts (system and custom).",
```
if this also includes system fonts, why is it called `getCustomFonts`?  wouldn't something like `getActiveFonts` or `getKnownFonts` be better?

also if this includes system fonts do we still need `getSupportedSystemFontFamilyNames`?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -122,6 +123,36 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+        if (!this.nodeStyles.customFonts.length) {
+            this._customFontsRow.showEmptyMessage();
+            this._customFontsRow.dataGrid = null;
+            return;
```
we typically dont add early `return` in functions like this as it makes it harder for new things to be added at the end in the future.  as such, id put everything after this in an `} else { ... }` (or more ideally flip it around and have the positive check be first with the error case in the `else`)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -220,10 +251,80 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+            }
+        };
+
+        this._dataGrid = new WI.DataGrid(fontColumns);
```
this needs a more unique name like `_allFontsDataGrid`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -220,10 +251,80 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+        };
+
+        this._dataGrid = new WI.DataGrid(fontColumns);
+        this._dataGrid.inline = true;
```
im not sure this is correct

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -220,10 +251,80 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+    }
+
+
+    _sortDataGrid()
```
ditto with a more unique name

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -220,10 +251,80 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+
+    _sortDataGrid()
+    {
+        var sortColumnIdentifier = this._dataGrid.sortColumnIdentifier;
```
Style: please use `let`

ditto below

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -220,10 +251,80 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+                title : WI.UIString("Family"),
+                sortable: true,
+                aligned: "centered",
+                width: "50%"
```
Style: missing trailing `,`

ditto below

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -220,10 +251,80 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+    _createDataGrid()
+    {
+        // All Fonts Properties
+        let fontColumns = {
```
Style: this can be `const` since it doesn't change between Web Inspector sessions

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -122,6 +123,36 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+        // with no custom fonts. This will trigger it to force
+        // rebuild the table.
+        if (!this._customFontsRow.dataGrid)
+            this._customFontsRow.dataGrid = this._dataGrid;
```
rather than keeping the `_allFontsDataGrid` as a member variable, i think instead we should `return` it from `_createAllFontsDataGrid` and then have that be called here
```js
this._customFontsRow.dataGrid ||= this._createAllFontsDataGrid();
this._customFontsRow.dataGrid.removeChildren();
...
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -122,6 +123,36 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+        this._customFontsRow.dataGrid.removeChildren();
+
+        for (let customFont of this.nodeStyles.customFonts) {
+
```
Style: unnecessary whitespace

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -122,6 +123,36 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+        for (let customFont of this.nodeStyles.customFonts) {
+
+            // The backend can return empty font families, which we should ignore.
+            if (customFont.family === "")
```
should we be filtering this in the backend?  im not sure i see a benefit in sending a font to the frontend if there's nothing to display for the name

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -122,6 +123,36 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+            if (customFont.family === "")
+                continue;
+
+            let dataGridNode = new WI.DataGridNode({family: customFont.family, weight: customFont.weight, stretch: customFont.stretch, style: customFont.stretch, loaded: customFont.loaded});
```
Style: please separate each property onto a separate line for better readability

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -122,6 +123,36 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+
+        // Custom Font Properties
+
+        // Display message when we have no custom fonts.
```
Style: these kinds of comments are not really useful as it's pretty straightforward from reading the code what this will do

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -340,6 +344,15 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
+            target.CSSAgent.getCustomFonts((error, fontFamilyNames) =>{
+                if (error)
+                    return;
+                this._customFonts = fontFamilyNames;
```
i think this should dispatch some sort of `WI.CSSManager.Event.FontsChanged` so that listeners (e.g. `WI.FontDetailsPanel`) can listen to it and update accordingly

dcrousso commented on Source/JavaScriptCore/inspector/protocol/CSS.json:
```diff
@@ -268,6 +268,21 @@
+              { "name": "featureSettings", "type": "string", "description": "Font Feature Settings"},
+              { "name": "display", "type": "string", "description": "Font Display"},
+              { "name": "sizeAdjust", "type": "string", "description": "Font Size Adjust"},
+              { "name": "loaded", "type": "string", "description": "Font Loaded"}
```
is this actually a useful piece of information for the developer?

are there any actions they can take given this info?  or is it purely informative?

if the latter, id suggest not including it as Web Inspector usually tries to only provide information that is actionable, as otherwise it's potentially confusing

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -916,6 +918,54 @@ Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<String>>> InspectorCSSAgent
+    auto customFonts = JSON::ArrayOf<Inspector::Protocol::CSS::FontFace>::create();
+    auto* domAgent = m_instrumentingAgents.persistentDOMAgent();
+
+    for (auto* document : domAgent->documents()) {
```
im not entirely sure what benefit there is in grabbing fonts for _all_ documents, as i dont think fonts are usable in situations like across `<iframe>`

perhaps we should adjust this to be something that gets the lists of fonts given a node (or more ideally a document node)?  that way we can show only the list of usable fonts for the selected node in the Fonts details sidebar panel

dcrousso reviewed (changes requested):
overall the code itself looks mostly fine with a few bugs and stylistic things to fix

conceptuatlly, im a bit concerned with the idea of this feature, as i think all of the other information in the Font details sidebar panel is related to the selected node, not general to the entire page (i.e. it could confuse the developer into incorrectly thinking that all the listed fonts are being used by the selected node)

also, given the fact that `document.fonts` exists in JS, im not sure what this provides to the developer beyond them evaluating that in the Console themselves (i.e. Web Inspector aims to provide unique functionality/information/etc. that cannot be accessed elsewhere)

id much rather see us be able to split the list of all fonts into ones used by the selected node (and ideally with the specific unicode ranges being used from the font) and those that aren't, as that way the developer can both understand what they are using as well as what it's possible to use (which could be fed into `font`/`font-family` autocomplete as well)

francescorn committed:
Web Inspector: Create All Fonts Tab in Font Section
https://bugs.webkit.org/show_bug.cgi?id=260127

Reviewed by NOBODY (OOPS!).

Co-authored by Brandon Stewart.

We created a Custom Fonts tab to display all of the Custom Fonts in the DOM, and to auto refresh to account for adjusting webpages such as slideshows.
There are parameters font-family name, stretch, style, and loaded, and can sort by each parameter as well.
If there are no custom fonts in the DOM, "No Custom Fonts" will be displayed.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getCustomFonts):
* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles):
(WI.DOMNodeStyles.prototype.get customFonts):
(WI.DOMNodeStyles.prototype.refresh):
* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel.prototype.update):
(WI.FontDetailsPanel.prototype.initialLayout):
(WI.FontDetailsPanel.prototype._sortDataGrid.comparator):
(WI.FontDetailsPanel.prototype._sortDataGrid):
